File: apps/desktop/src-tauri/src/commands/git/tests/fixtures/git_port.rs
use std::{
    path::Path,
    sync::{Arc, Mutex},
};

use anyhow::anyhow;
use host_domain::GitPort;
use host_domain::{
    GitAheadBehind, GitBranch, GitCommitAllRequest, GitConflictAbortRequest,
    GitConflictAbortResult, GitCurrentBranch, GitDiffScope, GitFileDiff, GitFileStatus,
    GitMergeBranchRequest, GitMergeBranchResult, GitPullRequest, GitPullResult, GitPushResult,
    GitRebaseAbortRequest, GitRebaseAbortResult, GitRebaseBranchRequest, GitRebaseBranchResult,
    GitResetWorktreeSelection, GitResetWorktreeSelectionRequest, GitResetWorktreeSelectionResult,
    GitWorktreeStatusData, GitWorktreeStatusSummaryData,
};

#[derive(Clone)]
pub(crate) enum WorktreeStatusResult {
    Ok(GitWorktreeStatusData),
    Err(String),
}

#[derive(Clone)]
pub(crate) enum WorktreeStatusSummaryResult {
    Ok(GitWorktreeStatusSummaryData),
    Err(String),
}

#[derive(Clone)]
pub(crate) enum ResetWorktreeSelectionResult {
    Ok(GitResetWorktreeSelectionResult),
    Err(String),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct WorktreeStatusCall {
    pub(crate) repo_path: String,
    pub(crate) target_branch: String,
    pub(crate) diff_scope: GitDiffScope,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct WorktreeStatusSummaryCall {
    pub(crate) repo_path: String,
    pub(crate) target_branch: String,
    pub(crate) diff_scope: GitDiffScope,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct CreateWorktreeCall {
    pub(crate) repo_path: String,
    pub(crate) worktree_path: String,
    pub(crate) branch: String,
    pub(crate) create_branch: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct RemoveWorktreeCall {
    pub(crate) repo_path: String,
    pub(crate) worktree_path: String,
    pub(crate) force: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct ResetWorktreeSelectionCall {
    pub(crate) repo_path: String,
    pub(crate) working_dir: Option<String>,
    pub(crate) target_branch: String,
    pub(crate) selection: GitResetWorktreeSelection,
}

pub(crate) struct CommandGitPortState {
    pub(crate) worktree_status_result: WorktreeStatusResult,
    pub(crate) worktree_status_calls: Vec<WorktreeStatusCall>,
    pub(crate) worktree_status_summary_result: WorktreeStatusSummaryResult,
    pub(crate) worktree_status_summary_calls: Vec<WorktreeStatusSummaryCall>,
    pub(crate) reset_worktree_selection_result: ResetWorktreeSelectionResult,
    pub(crate) reset_worktree_selection_calls: Vec<ResetWorktreeSelectionCall>,
    pub(crate) worktree_mutation_allowed: bool,
    pub(crate) create_worktree_calls: Vec<CreateWorktreeCall>,
    pub(crate) remove_worktree_calls: Vec<RemoveWorktreeCall>,
}

pub(crate) struct CommandGitPort {
    pub(crate) state: Arc<Mutex<CommandGitPortState>>,
}

impl CommandGitPort {
    pub(crate) fn new_with_summary_result(
        result: WorktreeStatusResult,
        summary_result: WorktreeStatusSummaryResult,
        worktree_mutation_allowed: bool,
    ) -> Self {
        Self {
            state: Arc::new(Mutex::new(CommandGitPortState {
                worktree_status_result: result,
                worktree_status_calls: Vec::new(),
                worktree_status_summary_result: summary_result,
                worktree_status_summary_calls: Vec::new(),
                reset_worktree_selection_result: ResetWorktreeSelectionResult::Ok(
                    GitResetWorktreeSelectionResult {
                        affected_paths: vec!["src/main.rs".to_string()],
                    },
                ),
                reset_worktree_selection_calls: Vec::new(),
                worktree_mutation_allowed,
                create_worktree_calls: Vec::new(),
                remove_worktree_calls: Vec::new(),
            })),
        }
    }
}

impl GitPort for CommandGitPort {
    fn get_branches(&self, _repo_path: &Path) -> anyhow::Result<Vec<GitBranch>> {
        panic!("unexpected call: get_branches");
    }

    fn get_current_branch(&self, _repo_path: &Path) -> anyhow::Result<GitCurrentBranch> {
        panic!("unexpected call: get_current_branch");
    }

    fn switch_branch(
        &self,
        _repo_path: &Path,
        _branch: &str,
        _create: bool,
    ) -> anyhow::Result<GitCurrentBranch> {
        panic!("unexpected call: switch_branch");
    }

    fn create_worktree(
        &self,
        repo_path: &Path,
        worktree_path: &Path,
        branch: &str,
        create_branch: bool,
    ) -> anyhow::Result<()> {
        let mut state = self
            .state
            .lock()
            .expect("command git port state lock should not be poisoned");
        if !state.worktree_mutation_allowed {
            panic!("unexpected call: create_worktree");
        }
        state.create_worktree_calls.push(CreateWorktreeCall {
            repo_path: repo_path.to_string_lossy().to_string(),
            worktree_path: worktree_path.to_string_lossy().to_string(),
            branch: branch.to_string(),
            create_branch,
        });
        Ok(())
    }

    fn remove_worktree(
        &self,
        repo_path: &Path,
        worktree_path: &Path,
        force: bool,
    ) -> anyhow::Result<()> {
        let mut state = self
            .state
            .lock()
            .expect("command git port state lock should not be poisoned");
        if !state.worktree_mutation_allowed {
            panic!("unexpected call: remove_worktree");
        }
        state.remove_worktree_calls.push(RemoveWorktreeCall {
            repo_path: repo_path.to_string_lossy().to_string(),
            worktree_path: worktree_path.to_string_lossy().to_string(),
            force,
        });
        Ok(())
    }

    fn delete_local_branch(
        &self,
        _repo_path: &Path,
        _branch: &str,
        _force: bool,
    ) -> anyhow::Result<()> {
        panic!("unexpected call: delete_local_branch");
    }

    fn push_branch(
        &self,
        _repo_path: &Path,
        _remote: &str,
        _branch: &str,
        _set_upstream: bool,
        _force_with_lease: bool,
    ) -> anyhow::Result<GitPushResult> {
        panic!("unexpected call: push_branch");
    }

    fn pull_branch(
        &self,
        _repo_path: &Path,
        _request: GitPullRequest,
    ) -> anyhow::Result<GitPullResult> {
        panic!("unexpected call: pull_branch");
    }

    fn rebase_abort(
        &self,
        _repo_path: &Path,
        _request: GitRebaseAbortRequest,
    ) -> anyhow::Result<GitRebaseAbortResult> {
        panic!("unexpected call: rebase_abort");
    }

    fn abort_conflict(
        &self,
        _repo_path: &Path,
        _request: GitConflictAbortRequest,
    ) -> anyhow::Result<GitConflictAbortResult> {
        panic!("unexpected call: abort_conflict");
    }

    fn merge_branch(
        &self,
        _repo_path: &Path,
        _request: GitMergeBranchRequest,
    ) -> anyhow::Result<GitMergeBranchResult> {
        panic!("unexpected call: merge_branch");
    }

    fn get_status(&self, _repo_path: &Path) -> anyhow::Result<Vec<GitFileStatus>> {
        panic!("unexpected call: get_status");
    }

    fn get_diff(
        &self,
        _repo_path: &Path,
        _target_branch: Option<&str>,
    ) -> anyhow::Result<Vec<GitFileDiff>> {
        panic!("unexpected call: get_diff");
    }

    fn get_worktree_status(
        &self,
        repo_path: &Path,
        target_branch: &str,
        diff_scope: GitDiffScope,
    ) -> anyhow::Result<GitWorktreeStatusData> {
        let mut state = self
            .state
            .lock()
            .expect("command git port state lock should not be poisoned");
        state.worktree_status_calls.push(WorktreeStatusCall {
            repo_path: repo_path.to_string_lossy().to_string(),
            target_branch: target_branch.to_string(),
            diff_scope,
        });
        match state.worktree_status_result.clone() {
            WorktreeStatusResult::Ok(payload) => Ok(payload),
            WorktreeStatusResult::Err(message) => Err(anyhow!(message)),
        }
    }

    fn get_worktree_status_summary(
        &self,
        repo_path: &Path,
        target_branch: &str,
        diff_scope: GitDiffScope,
    ) -> anyhow::Result<GitWorktreeStatusSummaryData> {
        let mut state = self
            .state
            .lock()
            .expect("command git port state lock should not be poisoned");
        state
            .worktree_status_summary_calls
            .push(WorktreeStatusSummaryCall {
                repo_path: repo_path.to_string_lossy().to_string(),
                target_branch: target_branch.to_string(),
                diff_scope,
            });
        match state.worktree_status_summary_result.clone() {
            WorktreeStatusSummaryResult::Ok(payload) => Ok(payload),
            WorktreeStatusSummaryResult::Err(message) => Err(anyhow!(message)),
        }
    }

    fn resolve_upstream_target(&self, _repo_path: &Path) -> anyhow::Result<Option<String>> {
        panic!("unexpected call: resolve_upstream_target");
    }

    fn suggested_squash_commit_message(
        &self,
        _repo_path: &Path,
        _source_branch: &str,
        _target_branch: &str,
    ) -> anyhow::Result<Option<String>> {
        panic!("unexpected call: suggested_squash_commit_message");
    }

    fn is_ancestor(
        &self,
        _repo_path: &Path,
        _ancestor_ref: &str,
        _descendant_ref: &str,
    ) -> anyhow::Result<bool> {
        panic!("unexpected call: is_ancestor");
    }

    fn commits_ahead_behind(
        &self,
        _repo_path: &Path,
        _target_branch: &str,
    ) -> anyhow::Result<GitAheadBehind> {
        panic!("unexpected call: commits_ahead_behind");
    }

    fn commit_all(
        &self,
        _repo_path: &Path,
        _request: GitCommitAllRequest,
    ) -> anyhow::Result<host_domain::GitCommitAllResult> {
        panic!("unexpected call: commit_all");
    }

    fn reset_worktree_selection(
        &self,
        repo_path: &Path,
        request: GitResetWorktreeSelectionRequest,
    ) -> anyhow::Result<GitResetWorktreeSelectionResult> {
        let mut state = self
            .state
            .lock()
            .expect("command git port state lock should not be poisoned");
        state
            .reset_worktree_selection_calls
            .push(ResetWorktreeSelectionCall {
                repo_path: repo_path.to_string_lossy().to_string(),
                working_dir: request.working_dir,
                target_branch: request.target_branch,
                selection: request.selection,
            });
        match state.reset_worktree_selection_result.clone() {
            ResetWorktreeSelectionResult::Ok(payload) => Ok(payload),
            ResetWorktreeSelectionResult::Err(message) => Err(anyhow!(message)),
        }
    }

    fn rebase_branch(
        &self,
        _repo_path: &Path,
        _request: GitRebaseBranchRequest,
    ) -> anyhow::Result<GitRebaseBranchResult> {
        panic!("unexpected call: rebase_branch");
    }
}

File: packages/host/src/ports/git-port.ts
import type {
  CommitsAheadBehind,
  FileDiff,
  FileStatus,
  GitBranch,
  GitCommitAllResult,
  GitConflict,
  GitConflictAbortResult,
  GitConflictOperation,
  GitCurrentBranch,
  GitDiffScope,
  GitFetchRemoteResult,
  GitFileStatusCounts,
  GitMergeMethod,
  GitPullBranchResult,
  GitPushBranchResult,
  GitRebaseAbortResult,
  GitRebaseBranchResult,
  GitResetWorktreeSelection,
  GitResetWorktreeSelectionResult,
  GitUpstreamAheadBehind,
} from "@openducktor/contracts";
import { Context, type Effect } from "effect";
import type {
  HostOperationError,
  HostPathAccessError,
  HostResourceError,
  HostValidationError,
} from "../effect/host-errors";

export type GitPortError =
  | HostOperationError
  | HostPathAccessError
  | HostResourceError
  | HostValidationError;

export type GitRemote = {
  name: string;
  url: string;
};
export type GitChangedFile = {
  path: string;
  status: string;
};
export type GitWorktreeStatusData = {
  currentBranch: GitCurrentBranch;
  fileStatuses: FileStatus[];
  fileDiffs: FileDiff[];
  targetAheadBehind: CommitsAheadBehind;
  upstreamAheadBehind: GitUpstreamAheadBehind;
  gitConflict?: GitConflict;
};
export type GitWorktreeStatusSummaryData = {
  currentBranch: GitCurrentBranch;
  fileStatuses: FileStatus[];
  fileStatusCounts: GitFileStatusCounts;
  targetAheadBehind: CommitsAheadBehind;
  upstreamAheadBehind: GitUpstreamAheadBehind;
  gitConflict?: GitConflict;
};
export type GitPushBranchOptions = {
  remote?: string;
  setUpstream?: boolean;
  forceWithLease?: boolean;
};
export type GitMergeBranchRequest = {
  sourceBranch: string;
  targetBranch: string;
  sourceWorkingDirectory?: string;
  method: GitMergeMethod;
  squashCommitMessage?: string;
};
export type GitMergeBranchResult =
  | {
      outcome: "merged";
      output: string;
    }
  | {
      outcome: "up_to_date";
      output: string;
    }
  | {
      outcome: "conflicts";
      conflictedFiles: string[];
      output: string;
    };
export type GitBranchUpstreamSetup = {
  createdTrackingRef: string | null;
};
export type GitPort = {
  canonicalizePath(path: string): Effect.Effect<string, HostOperationError>;
  isGitRepository(path: string): Effect.Effect<boolean, GitPortError>;
  shareGitCommonDirectory(
    repoPath: string,
    workingDir: string,
  ): Effect.Effect<boolean, GitPortError>;
  referenceExists(workingDir: string, reference: string): Effect.Effect<boolean, GitPortError>;
  listRemotes(workingDir: string): Effect.Effect<GitRemote[], GitPortError>;
  listBranches(workingDir: string): Effect.Effect<GitBranch[], GitPortError>;
  listFiles(workingDir: string): Effect.Effect<string[], GitPortError>;
  getCurrentBranch(workingDir: string): Effect.Effect<GitCurrentBranch, GitPortError>;
  getStatus(workingDir: string): Effect.Effect<FileStatus[], GitPortError>;
  listChangedFiles(
    workingDir: string,
    targetBranch: string,
  ): Effect.Effect<GitChangedFile[], GitPortError>;
  getDiff(workingDir: string, targetBranch?: string): Effect.Effect<FileDiff[], GitPortError>;
  getWorktreeStatusData(
    workingDir: string,
    targetBranch: string,
    diffScope: GitDiffScope,
  ): Effect.Effect<GitWorktreeStatusData, GitPortError>;
  getWorktreeStatusSummaryData(
    workingDir: string,
    targetBranch: string,
    diffScope: GitDiffScope,
  ): Effect.Effect<GitWorktreeStatusSummaryData, GitPortError>;
  createWorktree(
    repoPath: string,
    worktreePath: string,
    branch: string,
    createBranch: boolean,
    startPoint?: string,
  ): Effect.Effect<void, GitPortError>;
  configureBranchUpstream(
    repoPath: string,
    worktreePath: string,
    branch: string,
    upstreamRemote: string,
  ): Effect.Effect<GitBranchUpstreamSetup, GitPortError>;
  deleteReference(repoPath: string, reference: string): Effect.Effect<void, GitPortError>;
  removeWorktree(
    repoPath: string,
    worktreePath: string,
    force: boolean,
  ): Effect.Effect<void, GitPortError>;
  deleteLocalBranch(
    repoPath: string,
    branch: string,
    force: boolean,
  ): Effect.Effect<void, GitPortError>;
  isAncestor(
    workingDir: string,
    ancestor: string,
    descendant: string,
  ): Effect.Effect<boolean, GitPortError>;
  suggestedSquashCommitMessage(
    workingDir: string,
    sourceBranch: string,
    targetBranch: string,
  ): Effect.Effect<string | undefined, GitPortError>;
  mergeBranch(
    workingDir: string,
    request: GitMergeBranchRequest,
  ): Effect.Effect<GitMergeBranchResult, GitPortError>;
  switchBranch(
    workingDir: string,
    branch: string,
    create: boolean,
  ): Effect.Effect<GitCurrentBranch, GitPortError>;
  resetWorktreeSelection(
    workingDir: string,
    fileDiffs: FileDiff[],
    selection: GitResetWorktreeSelection,
  ): Effect.Effect<GitResetWorktreeSelectionResult, GitPortError>;
  commitsAheadBehind(
    workingDir: string,
    targetBranch: string,
  ): Effect.Effect<CommitsAheadBehind, GitPortError>;
  fetchRemote(
    workingDir: string,
    targetBranch: string,
  ): Effect.Effect<GitFetchRemoteResult, GitPortError>;
  pullBranch(workingDir: string): Effect.Effect<GitPullBranchResult, GitPortError>;
  commitAll(workingDir: string, message: string): Effect.Effect<GitCommitAllResult, GitPortError>;
  pushBranch(
    workingDir: string,
    branch: string,
    options?: GitPushBranchOptions,
  ): Effect.Effect<GitPushBranchResult, GitPortError>;
  rebaseBranch(
    workingDir: string,
    targetBranch: string,
  ): Effect.Effect<GitRebaseBranchResult, GitPortError>;
  rebaseAbort(workingDir: string): Effect.Effect<GitRebaseAbortResult, GitPortError>;
  abortConflict(
    workingDir: string,
    operation: GitConflictOperation,
  ): Effect.Effect<GitConflictAbortResult, GitPortError>;
};

export class GitPortTag extends Context.Tag("@openducktor/host/GitPort")<GitPortTag, GitPort>() {}
